Implemented Ellipsis in slicing Sympy Matrices.

arihantparsoya opened the pull request:
This is fix for  #11345. Added conditions when the argument is `ellipsis`.
#### Sample Program

``` python
>>> from sympy import *
>>> from sympy.abc import m, n
>>> X = MatrixSymbol('X', n, m)
>>> X[...]
X[:n, :m]
>>> X[2, ...]
X[2, :m]
>>> X[..., 2]
X[:n, 2]
```

arihantparsoya committed:
Implemented Ellipsis in sympy Matrices.

fix for  #11345.

Added test cases for Ellipsis arguments for matrix slicing.
… means the same as : in a matrix. Hence, if the argument is ‘….’ ,
input to MatrixSlice is set to `slice(None, None, None) `.

parsoyaarihant commented:
@leosartaj , can your review this?

parsoyaarihant commented:
@asmeurer , can you review this?

leosartaj commented:
This does not work.

```
>>> m = Matrix([[1, 2, 3], [4, 5, 6]])
>>> m[...]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-5-0ccb1cae6e36> in <module>()
----> 1 m[...]

/home/leosartaj/anaconda/lib/python2.7/site-packages/sympy/matrices/dense.pyc in __getitem__(self, key)
     94             if isinstance(key, slice):
     95                 return self._mat[key]
---> 96             return self._mat[a2idx(key)]
     97 
     98     def __setitem__(self, key, value):

/home/leosartaj/anaconda/lib/python2.7/site-packages/sympy/matrices/matrices.pyc in a2idx(j, n)
   4412             j = j.__index__()
   4413         except AttributeError:
-> 4414             raise IndexError("Invalid index a[%r]" % (j, ))
   4415     if n is not None:
   4416         if j < 0:

IndexError: Invalid index a[Ellipsis]
```

leosartaj commented:
This also doesn't work as expected.

```
>>> m = MatrixSymbol('A', 3, 4)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-16-0ccb1cae6e36> in <module>()
----> 1 m[...]

/home/leosartaj/anaconda/lib/python2.7/site-packages/sympy/matrices/expressions/matexpr.pyc in __getitem__(self, key)
    252                 raise IndexError("Single index only supported for "
    253                                  "non-symbolic indices.")
--> 254         raise IndexError("Invalid index, wanted %s[i,j]" % self)
    255 
    256     def as_explicit(self):

IndexError: Invalid index, wanted A[i,j]
```

I would expect something like this

```
>>> m[...]
A[:3, :4]
```

parsoyaarihant commented:
@leosartaj , the following program does work in my local computer:
```python
>>> from sympy import *
>>> m = MatrixSymbol('A', 3, 4)
>>> m[...]
A[:3, :4]
```

arihantparsoya committed:
Enabled slicing for DenseMatriX

… means the same as : in a matrix. Hence, if the argument is ‘….’ ,
input to MatrixSlice is set to `slice(None, None, None) `.

parsoyaarihant commented:
@leosartaj , the following program works now:

```python
>>> from sympy import *
>>> m = Matrix([[1, 2, 3], [4, 5, 6]])
>>> m[...]
Matrix([
[1, 2, 3],
[4, 5, 6]])
>>> m[0,...]
Matrix([[1, 2, 3]])
```

parsoyaarihant commented:
@leosartaj , ping.

parsoyaarihant commented:
@asmeurer , ping.

parsoyaarihant commented:
@asmeurer , ping.

bjodah commented on sympy/matrices/dense.py:
```diff
@@ -68,8 +68,16 @@ def __getitem__(self, key):
+                i = j = slice(None, None, None)
+            else:
+                i, j = key
+            if (i is Ellipsis) or (j is Ellipsis):
```
this line does not have any effect.

parsoyaarihant replied:
Thanks for reviewing this PR. If only one among `i` or `j` is `Ellipsis`, then this line will have use.
#### Example
```
assert X[2, ...] == MatrixSlice(X, 2, (0, m))
```

isuruf replied:
@bjodah is correct. the four lines below handle this.

parsoyaarihant replied:
Thanks, I have fixed it.

parsoyaarihant commented:
@leosartaj , ping.

isuruf commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -451,6 +451,9 @@ def test_slicing():
     m2 = Matrix([[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11], [12, 13, 14, 15]])
     assert m2[:, -1] == Matrix(4, 1, [3, 7, 11, 15])
     assert m2[-2:, :] == Matrix([[8, 9, 10, 11], [12, 13, 14, 15]])
+    assert m2[...] == m2
```
This isn't equal to `m2[:]` right? In other cases, `...` and `:` are equivalent.

parsoyaarihant replied:
No, its not equal.

sylee957 commented:
Although I don't see any clear advantage of slicing with ellipsis in matrix,
I see ellipsis slicing is implemented for numpy arrays and it can follow the same convention for sympy array.
So I think that this should be implemented first to sympy NDimArray and rather backported from there for sympy matrix.
